Boost: Trigger Cache Preloading on Page Cache activation

LiamSarsfield opened the pull request:
> [!NOTE]
> Merge this PR after https://github.com/Automattic/jetpack/pull/41988 is merged.

## Fixes
#42111

## Proposed changes:
* Schedules a preload of all Cornerstone Pages upon Boost's Page Cache activation.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pc9hqz-3nV-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
N/A

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

### Prerequisites
* Ensure that the constant `JETPACK_BOOST_ALPHA_FEATURES` is defined as true either in an mu-plugin or wp-config.php.
* If testing on a Jurassic Ninja site, ensure Edge Caching is disabled in Settings -> Edge Cache -> Disable Edge Cache
  * Bypassing Edge Cache is out of scope for this PR.
* Install [WP Crontrol plugin](https://wordpress.org/plugins/wp-crontrol/) 

### Testing
1. Go to Boost Settings page.
2. Ensure Page Cache is **not** enabled on the WordPress site.
3. Ensure Boost Premium is on the WordPress site.
4. Add at least 2 Cornerstone Pages to the list.
5. Delete the folder`./wp-content/boost-cache/cache/(domain)` within the WordPress site if it exists.
6. Enable Page Cache.

* **Before this applying this branch:**
  * Cornerstone Pages **will not** have cache files within `./wp-content/boost-cache/cache/(domain)`
* **After this applying this branch:**
  * Cornerstone Pages **will** have cache files within `./wp-content/boost-cache/cache/(domain)`

LiamSarsfield committed:
Page Cache: Add support for submodules with Cache Preload

Co-authored-by: Peter Petrov <dilirity@users.noreply.github.com>

LiamSarsfield committed:
changelog

LiamSarsfield committed:
Page Cache: Enhance Cache Preload with dynamic post scheduling and update mechanism

- Refactor preload mechanism to support dynamic post scheduling
- Add method to handle post updates and schedule preloading for cornerstone pages
- Implement more flexible preload post tracking using WordPress options
- Improve error handling during page preloading

LiamSarsfield committed:
Boost: Enhance Cache Preloading and Cornerstone Page Detection

- Add new method `is_cornerstone_page_by_url()` in Cornerstone_Utils to check cornerstone pages by URL
- Modify `is_cornerstone_page()` to use the new URL-based method
- Update Cache_Preload to handle cache invalidation and preloading of cornerstone pages
- Add action hook for successful cache invalidation
- Improve method names and add comments for clarity

LiamSarsfield committed:
Boost: Improve Cache Preload Handling for Cornerstone Pages

- Enhance `handle_cache_invalidation()` method to support different cache invalidation types
- Add support for full cache deletion by preloading all Cornerstone Pages
- Improve documentation with detailed comments and @since tags
- Update method signature to include invalidation type parameter
- Refactor cache preloading logic for more robust handling of cache events

LiamSarsfield committed:
Boost: Improve Cache Preload Performance and Error Handling

- Refactor page preloading to process URLs in batches of 10
- Add batch processing to reduce server load and improve reliability
- Enhance error handling with detailed logging for preload failures
- Improve HTTP response validation during page preloading
- Add methods to manage batch processing and track remaining URLs

LiamSarsfield committed:
Boost: Add Comprehensive Unit Tests for Cache Preload Module

- Create new test suite for Cache_Preload class in Jetpack Boost
- Implement comprehensive test coverage for cache preloading methods
- Test scenarios including cornerstone page preloading, cache invalidation, and scheduling
- Verify interactions with WordPress core functions and utility classes
- Ensure robust testing of preload scheduling, post update handling, and cache management

LiamSarsfield committed:
Boost: Update Cache Preload Test Suite with Direct Method Calls

- Refactor test cases to use direct method calls instead of partial mocks
- Replace Mockery-based mocking with explicit WordPress function expectations
- Improve test coverage by directly testing Cache_Preload class methods
- Simplify test setup and reduce complexity of test assertions
- Ensure consistent testing approach for cache preloading functionality

LiamSarsfield committed:
Boost: Add Cache-Busting Headers to Page Preload Requests

- Enhance page preloading by adding cache-control headers

LiamSarsfield committed:
Boost: Refactor Cache Preload Module with Feature Flag and Improvements

- Add feature flag support for cache preloading via JETPACK_BOOST_ALPHA_FEATURES constant
- Improve type hinting and method signatures in Cache_Preload class
- Update test suite to support new feature flag implementation

LiamSarsfield committed:
Boost: Implement Always-On Cache Preload and Improve Scheduling Logic

- Add `Is_Always_On` interface to Cache_Preload class
- Enhance cron job scheduling comment to explain 2-second delay rationale

LiamSarsfield committed:
Boost: Add Cache Preload Queue Management Class

- Implement `Cache_Preload_Queue_Manager` to handle preload queue operations
- Add methods for managing, adding, and processing preload URL batches
- Provide queue management functionality with clear, modular design

LiamSarsfield committed:
Boost: Optimize Cache Preload Queue Management

- Add in-memory caching for preload queue to reduce database calls
- Improve `get_posts_to_preload()` and `set_posts_to_preload()` methods
- Refactor `add_to_queue()` to use the new in-memory cache mechanism
- Simplify queue management with a single source of truth

LiamSarsfield committed:
Boost: Refactor Cache Preload Queue Manager with Batch Size Configuration

- Add constructor to set configurable batch size for preloading
- Update `prepare_next_batch()` method to use instance batch size
- Remove redundant batch size parameter from method signature

LiamSarsfield committed:
Enhance Cache Preload Documentation and Logic

- Update comments in Cache Preload class to clarify cache management processes
- Refactor conditional checks in File Storage class for improved readability and logic flow
- Ensure accurate descriptions of cache invalidation and rebuilding processes

LiamSarsfield committed:
Remove Cache Preload Queue Manager and Refactor Cache Preload Logic

- Delete Cache_Preload_Queue_Manager class to simplify preload management
- Refactor Cache_Preload class to directly handle preload queue operations
- Clear preload queue after processing to ensure accurate state management

LiamSarsfield committed:
Update Cache Preload Scheduling Logic to Remove Delay

- Adjusted the scheduling of the preload cron job to occur immediately instead of with a 2-second delay

LiamSarsfield committed:
Refactor Cache Preload Logic to Use Direct Page Parameter

- Removed unnecessary variable assignment for the page URL in the preload_page method

dilirity committed:
Fix My jetpack assets enqueued on every page instead of just Boost's

dilirity committed:
add changelog

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-cornerstone-pages

LiamSarsfield committed:
Refactor Cache Preload Methods to Accept Post URLs as Parameters

- Modified preload_pages to accept an array of posts as a parameter.
- Removed the old schedule_preload method and its associated logic.
- Adjusted handle_post_update to directly schedule preload for cornerstone pages.

LiamSarsfield committed:
Update Cache Preload to Use Array Parameters for Scheduling

- Refactored schedule_preload_cronjob to accept an array of posts.
- Adjusted tests to reflect changes in method signatures and ensure proper scheduling behavior.

LiamSarsfield committed:
Update test method names in Cache_Preload_Test to reflect cronjob functionality

- Renamed test methods to clarify that they are testing the schedule_preload_cronjob functionality.

LiamSarsfield committed:
Update Cache_Preload class to modify cronjob scheduling logic and streamline documentation

- Changed the return value of Cache Preloading feature check to false if not having features flag.
- Simplified comments for scheduling and preloading methods by removing unnecessary details.

LiamSarsfield committed:
Enhance Cache_Preload and Page_Cache classes with singleton pattern and activation logic

- Introduced a singleton pattern in Cache_Preload class with a private static instance and a public get_instance method.
- Updated Page_Cache class to implement Has_Activate interface and added an activate method to schedule cornerstone preload upon activation.

LiamSarsfield committed:
Add changelog entry for boost cache preloading feature activation

dilirity reviewed (dismissed):
Nice and simple!

One thing I'm worried about is ending up with all modules being singletons because we establish that pattern here 😅

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-on-cache-activation

LiamSarsfield committed:
Enhance Cache Preload Activation Logic

- Replace direct call to `Cache_Preload::get_instance()->schedule_cornerstone_preload()` with action hook `jetpack_boost_page_cache_activate` in the `activate` method of `Page_Cache` class.
- Remove singleton instance management from `Cache_Preload` class to simplify the preload logic and improve maintainability.
- Update the `setup` method in `Cache_Preload` to include the new action hook for scheduling cornerstone preloads on activation.

LiamSarsfield commented on projects/plugins/boost/app/modules/optimizations/page-cache/Page_Cache.php:
```diff
@@ -125,6 +126,10 @@ public function invalidate_cache() {
 	}
 
+	public static function activate() {
+		do_action( 'jetpack_boost_page_cache_activate' );
```
I've added the action here instead of calling `schedule_cornerstone_preload` directly as the feature flag was not taken into account when calling it directly. But the setup function that attaches the `jetpack_boost_page_cache_activate` with `schedule_cornerstone_preload` does as setup is not called if the feature flag is not enabled.

LiamSarsfield commented:
@dilirity Thanks for the review, I've updated the PR again to make it even simpler without needing the singleton pattern.

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/class-cache-preload.php:
```diff
@@ -54,19 +50,6 @@ public static function is_available() {
-		}
-
-		return self::$instance;
-	}
```
Fyi, you can use `Setup::get_instance_of( Cache_Preload::class )` to get the instance as long as it was already instantiated. @LiamSarsfield

LiamSarsfield replied:
> as long as it was already instantiated.

I've added a new function to ensure it's added to the `Setup` instance array already, as it isn't when it's needed upon the submodule activation.
https://github.com/Automattic/jetpack/blob/c41b018bfc58a1f4f72d74fcf45fcce230012473/projects/plugins/boost/app/lib/Setup.php#L55-L72

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/Page_Cache.php:
```diff
@@ -125,6 +126,10 @@ public function invalidate_cache() {
 	}
 
+	public static function activate() {
+		do_action( 'jetpack_boost_page_cache_activate' );
```
This doesn't actually work. This is because the submodule isn't loaded by the time the activation is being called for the cache module. But you could add an activation event on the submodule and preload from there.

LiamSarsfield replied:
> But you could add an activation event on the submodule and preload from there.

I was hesitant on doing this since the submodule is considered `Is_Always_On`. But technically the parent module isn't and can be "activated", hence the submodule is.
I've tested this locally and on JN, and the updated approach works now. @haqadn

haqadn reviewed (changes requested):
The functionality doesn't currently work. Check my [comment](https://github.com/Automattic/jetpack/pull/42528/files#r2001685795).

LiamSarsfield committed:
Add Cache Preload Activation Method and Update Class Implementations

- Introduced `get_or_create_instance_of` method in the `Setup` class to manage class instances more effectively.
- Updated `Cache_Preload` class to implement `Has_Activate` interface and added an `activate` method that utilizes the new instance management method for scheduling cornerstone preloads.
- Removed redundant `activate` method from `Page_Cache` class to streamline activation logic and ensure proper integration with the new preload scheduling mechanism.

LiamSarsfield committed:
Update Cache Preload Activation Documentation

- Added a comment to the `activate` method in the `Cache_Preload` class to clarify that activation is triggered by the parent module

LiamSarsfield committed:
Refactor `get_or_create_instance_of` Method in Setup Class

- Updated the method signature to enforce type hinting with `Has_Setup` for better clarity and type safety.
- Improved documentation to specify the expected class type for the parameter.

LiamSarsfield committed:
Remove redundant action hook for cornerstone preload scheduling in Cache_Preload setup method to streamline the activation process.

LiamSarsfield committed:
Update `get_or_create_instance_of` Method Signature in Setup Class

- Changed the parameter type hint from `Has_Setup` to `class-string<T>` for improved type safety and clarity in documentation.

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-on-cache-activation

LiamSarsfield committed:
Add comment to suppress Phan type mismatch warning in Setup class

- Added a comment to suppress the Phan type mismatch warning for the `add` method call in the `get_or_create_instance_of` method, ensuring clarity in type expectations.

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/class-cache-preload.php:
```diff
@@ -49,6 +51,16 @@ public static function is_available() {
+	 * @since $$next-version$$
+	 */
+	public static function activate() {
+		Setup::get_or_create_instance_of( self::class )->schedule_cornerstone_preload();
```
The advantage of having classes that have no side-effects during construction is that you do not always need the same instance. In this case, you can instantiate the class directly and skip adding the `Setup` call.

```suggestion
		$instance = new self();
		$instance->schedule_cornerstone_preload();
```

haqadn commented on projects/plugins/boost/app/lib/Setup.php:
```diff
@@ -51,4 +51,25 @@ public static function get_instance_of( $class_name ) {
 
 		return null;
 	}
+
+	/**
+	 * Get an instance of a class, or create a new one, and add it to the instances array if it doesn't exist.
+	 *
+	 * @template T of Has_Setup
+	 * @param class-string<T> $class_name
+	 * @return T
+	 */
+	public static function get_or_create_instance_of( $class_name ) {
+		$instance = self::get_instance_of( $class_name );
+		if ( $instance ) {
+			return $instance;
+		}
+
+		$instance = new $class_name();
+
+		// @phan-suppress-next-line PhanTypeMismatchArgument -- T implements Has_Setup per template definition
+		self::add( $instance );
+
+		return $instance;
+	}
```
We can create instance on demand as long as the constructor doesn't have side-effects. The reason we keep track of instances in `Setup` is because we want to be able to remove actions and filters, in which case having access to the instance is necessary. For example:

```php
remove_action( 'some_hook', [ Setup::get_instance_of( Page_Cache::class ), 'handle_hook' ] );
```

```suggestion
```

LiamSarsfield replied:
Makes sense, so if the class exists within `Setup`, it's safe to presume the `setup` would have ran, and `setup` side-effects can potentially be cleaned up as they are tracked within the `Setup` class.

haqadn replied:
Yes

haqadn reviewed (changes requested):
I have suggested a few changes. But, they are to keep the original intention of the existing design.

LiamSarsfield committed:
Refactor Cache Preload Activation in Cache_Preload Class

- Removed the use of `Setup::get_or_create_instance_of` in the `activate` method and replaced it with direct instantiation of the `Cache_Preload` class to simplify the activation process and improve clarity.

LiamSarsfield committed:
Merge branch 'trunk' into add/boost/cache-preloading-on-cache-activation

LiamSarsfield committed:
Merge branch 'trunk' into add/boost/cache-preloading-on-cache-activation

haqadn reviewed (approved):
Works as expected 👍🏼
